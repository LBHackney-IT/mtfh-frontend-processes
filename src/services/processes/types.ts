export enum Trigger {
  CloseProcess = "CloseProcess",
  CancelProcess = "CancelProcess",
}
