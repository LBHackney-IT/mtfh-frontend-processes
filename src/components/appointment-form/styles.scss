.mtfh-appointment-form__time {
  & .lbh-select {
    margin-top: 0;
  }
}
