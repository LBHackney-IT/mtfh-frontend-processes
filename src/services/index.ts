export { default as locale } from "./locale";
export { default as menu } from "./menu";
export * as processes from "./processes";
export * from "./config";
