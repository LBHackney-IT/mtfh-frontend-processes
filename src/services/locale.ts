import reactHtmlParser from "react-html-parser";

import { AssetAddress } from "@mtfh/common/lib/api/asset/v1";

const locale = {
  backButton: "Back",
  bookAppointment: "Book Appointment",
  next: "Next",
  cancel: "Cancel",
  cancelProcess: "Cancel process",
  change: "Change",
  views: {
    soleToJoint: {
      title: "Sole tenant requests a joint tenure",
      steps: {
        selectTenant: "Select person",
        checkEligibility: "Eligibility Checks",
        supportingDocuments: "Supporting documents",
        finish: "Finish",
        breachOfTenancy: "Breach of tenure check",
        requestDocuments: "Request Documents",
        reviewDocuments: "Review Documents",
        submitCase: "Submit case",
      },
      actions: {
        reassignCase: "Reassign case",
        cancelProcess: "Cancel process",
        caseActivityHistory: "Case activity history",
      },
    },
    selectTenants: {
      selectTenantHint:
        "This person will be asked for proof of relationship e.g. marriage or civil partnership certificate.",
      selectTenantLabel: "Who do you want to add as a joint tenant?",
      addToTenureText: "If the person you want to add is not listed you must first",
      addToTenureLink: "add them to the tenure",
      noHouseholdMembersOver18:
        "There are no household members over the age of 18 associated to this.",
    },
    processesMenu: {
      selectProcessDescription: () =>
        reactHtmlParser(
          "<p>Start a process by selecting a process category below. <br /> Where a process is not yet supported in <strong>Manage My Home</strong> these links will open the appropriate interim Google form.</p>",
        ),
    },
    checkEligibility: {
      autoCheckIntro:
        "Manage My Home has used tenure, person and property information to automatically check eligibility for this application.",
      passedChecks: "Passed automatic eligibilty checks",
      failedChecks: "Not eligible for a sole to joint tenure",
    },
    breachOfTenancy: {
      autoCheckIntro:
        "Manage My Home has used tenure, person and property information to automatically check eligibility for this application.",
      passedChecks: "Eligibility checks passed",
      failedChecks: "Not eligible for a sole to joint tenure",
    },
    reviewDocuments: {
      passedChecks: "Eligibility checks passed",
      documentsRequested: "Documents requested",
      checkSupportingDocumentsAppointment:
        "I need to make an appointment to check supporting documents",
      appointmentScheduled: "Office appointment scheduled",
      seenPhotographicId:
        "I confirm that an outcome letter has been sent to the resident",
      seenSecondId:
        "I confirm I have seen a second form of ID (does not have to be photographic)",
      isNotInImmigrationControl:
        "I confirm that the prospective tenant is not subject to immigration control under the Asylum And Immigration Act 1996",
      seenProofOfRelationship:
        "I confirm I have seen proof of relationship to the existing tenant",
      incomingTenantLivingInProperty:
        "I confirm I have seen 3 separate documents proving the proposed tenant has been living at the property for a minimum of 12 months",
    },
  },
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this type.",
    unableToFindState: "Invalid State",
    unableToFindStateDescription: "We have been unable to process a state of this type.",
  },
  title: "Processes",
  loadingText: "Submitting...",
  components: {
    startProcess: {
      buttonLabel: "Start process",
      thirdPartyHeading: "Sharing information with third parties",
    },
    entitySummary: {
      tenurePaymentRef: "Tenure payment ref",
      address: (assetAddress: AssetAddress): string => {
        const {
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,
          addressLine4,
          postCode,
        } = assetAddress;

        return [
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,

          addressLine4,
          postCode,
        ]
          .filter((addressLine) => !!addressLine)
          .join(" ");
      },
    },
  },
};

export default locale;
