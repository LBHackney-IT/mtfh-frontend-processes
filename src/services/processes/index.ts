export * from "./sole-to-joint";
export * from "./change-of-name";
