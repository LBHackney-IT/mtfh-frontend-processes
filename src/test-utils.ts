import {
  getAssetV1,
  getPersonV1,
  getProcessV1,
  getTenureV1,
  mockProcessV1,
  patchProcessV1,
  postProcessV1,
  server,
} from "@hackney/mtfh-test-utils";

beforeEach(() => {
  server.use(getTenureV1());
  server.use(getPersonV1());
  server.use(getAssetV1());
  server.use(postProcessV1());
  server.use(getProcessV1());
  server.use(patchProcessV1());
});

export const mockProcessSelectTenants = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "SelectTenants" },
};

export const mockProcessAutomatedChecksFailed = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "AutomatedChecksFailed" },
};

export const mockProcessAutomatedChecksPassed = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "AutomatedChecksPassed" },
};

export const mockProcessInvalidState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "InvalidState" },
};

export const mockManualChecksPassedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "ManualChecksPassed" },
};

export const mockManualChecksFailedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "ManualChecksFailed" },
};

export const mockBreachChecksFailedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "BreachChecksFailed" },
};

export const mockBreachChecksPassedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "BreachChecksPassed" },
};

export const mockDocumentsRequestedDes = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "DocumentsRequestedDes" },
};

export const mockProcessClosedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "ProcessClosed" },
};

export const mockDocumentsRequestedDesAppointment = {
  ...mockProcessV1,
  currentState: {
    ...mockProcessV1.currentState,
    state: "DocumentsRequestedAppointment",
    processData: {
      formData: {
        appointmentDateTime: "2022-10-12T08:59:00.000Z",
      },
      documents: [],
    },
  },
  previousStates: [{ ...mockProcessV1.currentState, state: "DocumentsRequestedDes" }],
};

export const mockDocumentsRequestedAppointment = ({
  appointmentDateTime,
}: {
  appointmentDateTime: string;
}) => {
  return {
    ...mockProcessV1,
    currentState: {
      ...mockProcessV1.currentState,
      state: "DocumentsRequestedAppointment",
      processData: {
        formData: {
          appointmentDateTime,
        },
        documents: [],
      },
    },
  };
};

export const mockDocumentsAppointmentRescheduled = {
  ...mockProcessV1,
  currentState: {
    ...mockProcessV1.currentState,
    state: "DocumentsAppointmentRescheduled",
  },
};
