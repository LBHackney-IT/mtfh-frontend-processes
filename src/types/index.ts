export type TargetType = "tenure" | "property" | "person";

export interface IStartProcess {
  thirdPartyCondition?: string;
  thirdPartyComponent?: React.FC;
  subHeading?: string;
  subComponent?: React.FC;
}

export interface IProcess {
  processName: string;
  title: string;
  targetType: TargetType;
  startProcess: IStartProcess;
  states: {
    [state: string]: {
      state: string;
      triggers: Record<string, string>;
    };
  };
}
