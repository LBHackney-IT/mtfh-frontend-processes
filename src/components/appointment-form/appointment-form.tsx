import React from "react";

import { isPast } from "date-fns";
import { Form, Formik } from "formik";

import {
  AppointmentFormData,
  appointmentSchema,
  dateTimeIsValid,
} from "../../schemas/appointment";
import { locale } from "../../services";
import { getAppointmentDateTime } from "../../views/process-view/sole-to-joint-view/states/shared";

import { Process, editProcess } from "@mtfh/common/lib/api/process/v1";
import {
  Button,
  Center,
  DateField,
  FormGroup,
  Spinner,
  TimeField,
} from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

import "./styles.scss";

interface BookAppointmentFormProps {
  process: Process;
  mutate: () => void;
  needAppointment: boolean;
  setGlobalError: any;
  setNeedAppointment?: any;
  options: {
    buttonText?: string;
    requestAppointmentTrigger: string;
    rescheduleAppointmentTrigger: string;
    appointmentRequestedState: string;
    appointmentRescheduledState: string;
  };
}

export const AppointmentForm = ({
  process,
  mutate,
  setGlobalError,
  needAppointment,
  setNeedAppointment,
  options,
}: BookAppointmentFormProps): JSX.Element => {
  const { currentState } = process;
  const formData = process.currentState.processData.formData as {
    appointmentDateTime: string;
  };
  const errorMessages = useErrorCodes();

  if (!errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }
  return (
    <>
      <Formik<AppointmentFormData>
        initialValues={{ day: "", month: "", year: "", hour: "", minute: "", amPm: "" }}
        validateOnBlur={false}
        validateOnChange={false}
        validationSchema={appointmentSchema(errorMessages)}
        validate={(values) => {
          return validate(errorMessages, values);
        }}
        onSubmit={async (values) => {
          const appointmentDateTime = getAppointmentDateTime(values);
          let processTrigger = options.requestAppointmentTrigger;
          if (
            [
              options.appointmentRequestedState,
              options.appointmentRescheduledState,
            ].includes(currentState.state)
          ) {
            if (isPast(new Date(formData.appointmentDateTime))) {
              processTrigger = options.rescheduleAppointmentTrigger;
            } else {
              processTrigger = "";
            }
          }
          try {
            await editProcess({
              id: process.id,
              processTrigger,
              processName: process?.processName,
              etag: process.etag || "",
              formData: {
                appointmentDateTime,
              },
              documents: [],
              processData: {
                formData: {
                  appointmentDateTime,
                },
                documents: [],
              },
            });
            setNeedAppointment(!needAppointment);
            mutate();
          } catch (e: any) {
            setGlobalError(e.response?.status || 500);
          }
        }}
      >
        {({ values, errors }) => {
          return (
            needAppointment && (
              <Form
                noValidate
                id="request-appointment-form"
                className="request-appointment-form"
              >
                <DateTimeFields errors={errors} />
                <Button
                  type="submit"
                  disabled={
                    !Object.values(values).some((value) => {
                      return value !== "";
                    })
                  }
                  style={{ width: 222 }}
                >
                  {options.buttonText || locale.confirm}
                </Button>
              </Form>
            )
          );
        }}
      </Formik>
    </>
  );
};

export const validate = (errorMessages, values) => {
  if (
    values.year ||
    values.month ||
    values.day ||
    values.hour ||
    values.minute ||
    values.amPm
  ) {
    const result = dateTimeIsValid(errorMessages, values);
    if (result === true) {
      return {};
    }
    return { global: result };
  }
};

export const DateTimeFields = ({
  dateLabel = "Date",
  timeLabel = "Time",
  errors,
}: {
  dateLabel?: string;
  timeLabel?: string;
  errors?: any;
}): JSX.Element => {
  return (
    <FormGroup id="appointment-form-date-time-form-group" error={errors?.global}>
      <FormGroup
        id="appointment-form-date-time-field-form-group"
        style={{
          display: "flex",
          flexWrap: "wrap",
          alignItems: "flex-end",
          marginTop: 0,
        }}
      >
        <>
          <DateField
            id="appointment-form-date"
            className="mtfh-appointment-form__date"
            label={dateLabel}
            dayLabel=""
            monthLabel=""
            yearLabel=""
            dayProps={{ name: "day", placeholder: "dd" }}
            monthProps={{ name: "month", placeholder: "mm" }}
            yearProps={{ name: "year", placeholder: "yyyy" }}
            style={{ flex: "1 1 280px" }}
            required
          />
          <TimeField
            id="appointment-form-time"
            className="mtfh-appointment-form__time"
            label={timeLabel}
            hourLabel=""
            minuteLabel=""
            amPmLabel=""
            hourProps={{ name: "hour", placeholder: "00" }}
            minuteProps={{ name: "minute", placeholder: "00" }}
            amPmProps={{ name: "amPm" }}
            style={{ flex: "1 1 280px" }}
            required
          />
        </>
      </FormGroup>
    </FormGroup>
  );
};
