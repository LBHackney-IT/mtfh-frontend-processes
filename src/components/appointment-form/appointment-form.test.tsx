import React from "react";

import { getReferenceDataV1, render, server } from "@hackney/mtfh-test-utils";
import { screen, waitForElementToBeRemoved } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { processes } from "../../services";
import { Trigger } from "../../services/processes/types";
import { mockDocumentsRequestedDes } from "../../test-utils";
import { AppointmentForm } from "./appointment-form";

const setNeedAppointment = jest.fn();
describe("appointment-form-component", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  test("it renders appointment form correctly", async () => {
    server.use(getReferenceDataV1({}, 200));
    const { states } = processes.soletojoint;
    const { container } = render(
      <AppointmentForm
        process={mockDocumentsRequestedDes}
        mutate={() => {}}
        needAppointment
        setGlobalError={() => {}}
        setNeedAppointment={setNeedAppointment}
        appointmentTrigger=""
        options={{
          buttonText: "Continue",
          requestAppointmentTrigger: Trigger.RequestDocumentsAppointment,
          rescheduleAppointmentTrigger: Trigger.RescheduleDocumentsAppointment,
          appointmentRequestedState: states.documentsRequestedAppointment.state,
          appointmentRescheduledState: states.documentsAppointmentRescheduled.state,
        }}
      />,
    );
    await waitForElementToBeRemoved(screen.queryAllByText(/Loading/));
    expect(container).toMatchSnapshot();
    expect(screen.getByText("Continue")).toBeDisabled();
    await userEvent.type(screen.getByPlaceholderText(/yyyy/i), "2000");
    expect(screen.getByText("Continue")).toBeEnabled();
  });

  test("it doesnt render appointment form", async () => {
    server.use(getReferenceDataV1({}, 200));
    const { states } = processes.soletojoint;
    const { container } = render(
      <AppointmentForm
        process={mockDocumentsRequestedDes}
        mutate={() => {}}
        needAppointment={false}
        setGlobalError={() => {}}
        setNeedAppointment={setNeedAppointment}
        appointmentTrigger=""
        options={{
          buttonText: "Continue",
          requestAppointmentTrigger: Trigger.RequestDocumentsAppointment,
          rescheduleAppointmentTrigger: Trigger.RescheduleDocumentsAppointment,
          appointmentRequestedState: states.documentsRequestedAppointment.state,
          appointmentRescheduledState: states.documentsAppointmentRescheduled.state,
        }}
      />,
    );
    await waitForElementToBeRemoved(screen.queryAllByText(/Loading/));
    expect(container).toMatchSnapshot();
  });
});
