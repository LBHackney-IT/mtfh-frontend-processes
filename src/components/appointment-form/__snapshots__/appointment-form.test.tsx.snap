// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appointment-form-component it doesnt render appointment form 1`] = `<div />`;

exports[`appointment-form-component it renders appointment form correctly 1`] = `
<div>
  <form
    action="#"
    class="request-appointment-form"
    id="request-appointment-form"
    novalidate=""
  >
    <div
      class="govuk-form-group lbh-form-group"
      id="appointment-form-date-time-form-group"
    >
      <label
        class="govuk-label lbh-label"
        for="appointment-form-date-time-form-group-field"
      >
        
      </label>
      <div
        class="govuk-form-group lbh-form-group"
        id="appointment-form-date-time-field-form-group"
        style="display: flex; flex-wrap: wrap; align-items: flex-end; margin-top: 0px;"
      >
        <label
          class="govuk-label lbh-label"
          for="appointment-form-date-time-field-form-group-field"
        >
          
        </label>
        <fieldset
          class="govuk-form-group lbh-form-group mtfh-appointment-form__date"
          id="appointment-form-date"
          style="flex: 1 1 280px;"
        >
          <legend
            class="govuk-label lbh-label"
            for="appointment-form-date-field"
          >
            Date
            <sup
              aria-hidden="true"
            >
              *
            </sup>
          </legend>
          <div
            class="govuk-date-input lbh-date-input"
          >
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="appointment-form-date-field-day"
              />
              <input
                aria-label="Day"
                aria-valuemax="31"
                aria-valuemin="1"
                class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
                name="day"
                placeholder="dd"
                required=""
                role="spinbutton"
                value=""
              />
            </div>
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="appointment-form-date-field-month"
              />
              <input
                aria-label="Month"
                aria-valuemax="12"
                aria-valuemin="1"
                class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
                name="month"
                placeholder="mm"
                required=""
                role="spinbutton"
                value=""
              />
            </div>
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="appointment-form-date-field-year"
              />
              <input
                aria-label="Year"
                class="govuk-input lbh-input govuk-input govuk-date-input__input govuk-input--width-4"
                name="year"
                placeholder="yyyy"
                required=""
                role="spinbutton"
                value=""
              />
            </div>
          </div>
        </fieldset>
        <fieldset
          class="govuk-form-group lbh-form-group mtfh-appointment-form__time"
          id="appointment-form-time"
          style="flex: 1 1 280px;"
        >
          <legend
            class="govuk-label lbh-label"
            for="appointment-form-time-field"
          >
            Time
            <sup
              aria-hidden="true"
            >
              *
            </sup>
          </legend>
          <div
            class="govuk-date-input lbh-date-input"
          >
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="appointment-form-time-field-hour"
              />
              <input
                aria-label="Hour"
                aria-valuemax="12"
                aria-valuemin="0"
                class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
                name="hour"
                placeholder="00"
                required=""
                role="spinbutton"
                value=""
              />
            </div>
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="appointment-form-time-field-minute"
              />
              <input
                aria-label="Minute"
                aria-valuemax="59"
                aria-valuemin="0"
                class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
                name="minute"
                placeholder="00"
                required=""
                role="spinbutton"
                value=""
              />
            </div>
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="appointment-form-time-field-amPm"
              />
              <select
                aria-label="AM/PM"
                class="govuk-select lbh-select"
                id="amPm"
                name="amPm"
              >
                <option
                  value=""
                >
                  AM/PM
                </option>
                <option
                  value="am"
                >
                  AM
                </option>
                <option
                  value="pm"
                >
                  PM
                </option>
              </select>
            </div>
          </div>
        </fieldset>
      </div>
    </div>
    <button
      class="govuk-button lbh-button"
      disabled=""
      style="width: 222px;"
      type="submit"
    >
      Continue
    </button>
  </form>
</div>
`;
